RESTWS-938: Add an API to retrieve NameTemplate by id...

xprl-gjf opened the pull request:
...or to list all NameTemplates

## Description of what I changed

* Augmented `NameTemplateController2_0` to support returning a specific, named, NameTemplate instance (while maintaining the existing functionality to return the default NameTemplate configured via global property `layout.name.format`)
* Refactored `NameTemplateController2_0` and `AddressTemplateController2_0` by moving common functionality into a generic `LayoutTemplateProvider` class
* Added a new `NameTemplatesController2_0` for the new API endpoint to list all known NameTemplates.
* Added integration tests.

<!--- It can simply be your commit message, which you must have -->


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/RESTWS-938

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

xprl-gjf commented:
Examples of API output, from integration tests run on my own OpenMRS installation:

GET /openmrs/ws/rest/v1/nametemplate
reponse: 200
```json
{"displayName":"Formato de Nombres en Latino America","codeName":"latinamerica","country":null,"lines":[[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"First Name","codeName":"givenName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Second Name","codeName":"middleName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 1","codeName":"familyName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 2","codeName":"familyName2","displaySize":"30"}]],"lineByLineFormat":["givenName","middleName","familyName","familyName2"],"nameMappings":{"familyName2":"Family Name 2","givenName":"First Name","familyName":"Family Name 1","middleName":"Second Name"},"sizeMappings":{"familyName2":"30","givenName":"30","familyName":"30","middleName":"30"},"elementDefaults":null,"elementRegex":null,"elementRegexFormats":null,"requiredElements":null}
```

GET /openmrs/ws/rest/v1/nametemplate/global
response: 200
```json
{"displayName":"Formato de Nombres en Latino America","codeName":"latinamerica","country":null,"lines":[[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"First Name","codeName":"givenName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Second Name","codeName":"middleName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 1","codeName":"familyName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 2","codeName":"familyName2","displaySize":"30"}]],"lineByLineFormat":["givenName","middleName","familyName","familyName2"],"nameMappings":{"familyName2":"Family Name 2","givenName":"First Name","familyName":"Family Name 1","middleName":"Second Name"},"sizeMappings":{"familyName2":"30","givenName":"30","familyName":"30","middleName":"30"},"elementDefaults":null,"elementRegex":null,"elementRegexFormats":null,"requiredElements":null}
```

GET /openmrs/ws/rest/v1/nametemplate/short
response: 200
```json
{"displayName":"Short Name Format","codeName":"short","country":null,"lines":[[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Given","codeName":"givenName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Middle","codeName":"middleName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name","codeName":"familyName","displaySize":"30"}]],"lineByLineFormat":["givenName","middleName","familyName"],"nameMappings":{"givenName":"Given","familyName":"Family Name","middleName":"Middle"},"sizeMappings":{"givenName":"30","familyName":"30","middleName":"30"},"elementDefaults":null,"elementRegex":null,"elementRegexFormats":null,"requiredElements":null}
```

GET /openmrs/ws/rest/v1/nametemplate/long
response: 200
```json
{"displayName":"Long Name Format","codeName":"long","country":null,"lines":[[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Prefix","codeName":"prefix","displaySize":"5"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Given","codeName":"givenName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Middle","codeName":"middleName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name Prefix","codeName":"familyNamePrefix","displaySize":"8"},{"isToken":"IS_NOT_NAME_TOKEN","displayText":" "},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name","codeName":"familyName","displaySize":"25"},{"isToken":"IS_NOT_NAME_TOKEN","displayText":" "},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 2","codeName":"familyName2","displaySize":"25"},{"isToken":"IS_NOT_NAME_TOKEN","displayText":" "},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name Suffix","codeName":"familyNameSuffix","displaySize":"8"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Degree","codeName":"degree","displaySize":"5"}]],"lineByLineFormat":["prefix","givenName","middleName","familyNamePrefix familyName familyName2 familyNameSuffix","degree"],"nameMappings":{"familyName2":"Family Name 2","prefix":"Prefix","familyNamePrefix":"Family Name Prefix","givenName":"Given","familyName":"Family Name","degree":"Degree","middleName":"Middle","familyNameSuffix":"Family Name Suffix"},"sizeMappings":{"familyName2":"25","prefix":"5","familyNamePrefix":"8","givenName":"30","familyName":"25","degree":"5","middleName":"30","familyNameSuffix":"8"},"elementDefaults":null,"elementRegex":null,"elementRegexFormats":null,"requiredElements":null}
```

GET /openmrs/ws/rest/v1/nametemplate/latinamerica
reponse: 200
```json
{"displayName":"Formato de Nombres en Latino America","codeName":"latinamerica","country":null,"lines":[[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"First Name","codeName":"givenName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Second Name","codeName":"middleName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 1","codeName":"familyName","displaySize":"30"}],[{"isToken":"IS_NOT_NAME_TOKEN","displayText":""},{"isToken":"IS_NAME_TOKEN","displayText":"Family Name 2","codeName":"familyName2","displaySize":"30"}]],"lineByLineFormat":["givenName","middleName","familyName","familyName2"],"nameMappings":{"familyName2":"Family Name 2","givenName":"First Name","familyName":"Family Name 1","middleName":"Second Name"},"sizeMappings":{"familyName2":"30","givenName":"30","familyName":"30","middleName":"30"},"elementDefaults":null,"elementRegex":null,"elementRegexFormats":null,"requiredElements":null}
```

GET /openmrs/ws/rest/v1/nametemplate/foo
response: 404 Not Found

GET /openmrs/ws/rest/v1/nametemplates
response: 200
```json
{
  "results": [
    {
      "displayName": "Formato de Nombres en Espana",
      "codeName": "spain",
      "country": null,
      "lines": [
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Given",
            "codeName": "givenName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name",
            "codeName": "familyName",
            "displaySize": "25"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name 2",
            "codeName": "familyName2",
            "displaySize": "25"
          }
        ]
      ],
      "lineByLineFormat": [
        "givenName",
        "familyName",
        "familyName2"
      ],
      "nameMappings": {
        "familyName2": "Family Name 2",
        "givenName": "Given",
        "familyName": "Family Name"
      },
      "sizeMappings": {
        "familyName2": "25",
        "givenName": "30",
        "familyName": "25"
      },
      "elementDefaults": null,
      "elementRegex": null,
      "elementRegexFormats": null,
      "requiredElements": null
    },
    {
      "displayName": "Formato de Nombres en Latino America",
      "codeName": "latinamerica",
      "country": null,
      "lines": [
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "First Name",
            "codeName": "givenName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Second Name",
            "codeName": "middleName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name 1",
            "codeName": "familyName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name 2",
            "codeName": "familyName2",
            "displaySize": "30"
          }
        ]
      ],
      "lineByLineFormat": [
        "givenName",
        "middleName",
        "familyName",
        "familyName2"
      ],
      "nameMappings": {
        "familyName2": "Family Name 2",
        "givenName": "First Name",
        "familyName": "Family Name 1",
        "middleName": "Second Name"
      },
      "sizeMappings": {
        "familyName2": "30",
        "givenName": "30",
        "familyName": "30",
        "middleName": "30"
      },
      "elementDefaults": null,
      "elementRegex": null,
      "elementRegexFormats": null,
      "requiredElements": null
    },
    {
      "displayName": "\"Given Family\" Format",
      "codeName": "givenfamily",
      "country": null,
      "lines": [
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Given",
            "codeName": "givenName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name",
            "codeName": "familyName",
            "displaySize": "30"
          }
        ]
      ],
      "lineByLineFormat": [
        "givenName",
        "familyName"
      ],
      "nameMappings": {
        "givenName": "Given",
        "familyName": "Family Name"
      },
      "sizeMappings": {
        "givenName": "30",
        "familyName": "30"
      },
      "elementDefaults": null,
      "elementRegex": null,
      "elementRegexFormats": null,
      "requiredElements": null
    },
    {
      "displayName": "Short Name Format",
      "codeName": "short",
      "country": null,
      "lines": [
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Given",
            "codeName": "givenName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Middle",
            "codeName": "middleName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name",
            "codeName": "familyName",
            "displaySize": "30"
          }
        ]
      ],
      "lineByLineFormat": [
        "givenName",
        "middleName",
        "familyName"
      ],
      "nameMappings": {
        "givenName": "Given",
        "familyName": "Family Name",
        "middleName": "Middle"
      },
      "sizeMappings": {
        "givenName": "30",
        "familyName": "30",
        "middleName": "30"
      },
      "elementDefaults": null,
      "elementRegex": null,
      "elementRegexFormats": null,
      "requiredElements": null
    },
    {
      "displayName": "Long Name Format",
      "codeName": "long",
      "country": null,
      "lines": [
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Prefix",
            "codeName": "prefix",
            "displaySize": "5"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Given",
            "codeName": "givenName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Middle",
            "codeName": "middleName",
            "displaySize": "30"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name Prefix",
            "codeName": "familyNamePrefix",
            "displaySize": "8"
          },
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": " "
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name",
            "codeName": "familyName",
            "displaySize": "25"
          },
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": " "
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name 2",
            "codeName": "familyName2",
            "displaySize": "25"
          },
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": " "
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Family Name Suffix",
            "codeName": "familyNameSuffix",
            "displaySize": "8"
          }
        ],
        [
          {
            "isToken": "IS_NOT_NAME_TOKEN",
            "displayText": ""
          },
          {
            "isToken": "IS_NAME_TOKEN",
            "displayText": "Degree",
            "codeName": "degree",
            "displaySize": "5"
          }
        ]
      ],
      "lineByLineFormat": [
        "prefix",
        "givenName",
        "middleName",
        "familyNamePrefix familyName familyName2 familyNameSuffix",
        "degree"
      ],
      "nameMappings": {
        "familyName2": "Family Name 2",
        "prefix": "Prefix",
        "familyNamePrefix": "Family Name Prefix",
        "givenName": "Given",
        "familyName": "Family Name",
        "degree": "Degree",
        "middleName": "Middle",
        "familyNameSuffix": "Family Name Suffix"
      },
      "sizeMappings": {
        "familyName2": "25",
        "prefix": "5",
        "familyNamePrefix": "8",
        "givenName": "30",
        "familyName": "25",
        "degree": "5",
        "middleName": "30",
        "familyNameSuffix": "8"
      },
      "elementDefaults": null,
      "elementRegex": null,
      "elementRegexFormats": null,
      "requiredElements": null
    }
  ]
}
```

xprl-gjf commented:
@mogoodrich , @dkayiwa , any chance you could review this, or nominate somebody more appropriate?

dkayiwa commented:
@xprl-gjf shouldn't you just add parameters to the existing nametemplate resource?

xprl-gjf commented:
> @xprl-gjf shouldn't you just add parameters to the existing nametemplate resource?

There isn't an existing nametemplate resource in the sense of the normal conventions of the OpenMRS REST API. The existing API was implemented directly via a controller. Worse (in my opinion), the existing API does not conform to the conventions of a resource in that the existing API method /nametemplate does not list all nametemplates; instead it returns one specific nametemplate (the nametemplate specified by the global setting `layout.name.template`)

In effect, I have added a 'parameter' to the existing 'resource' by adding an API endpoint GET /nametemplate/{template-id}.

The complication comes with the endpoint I am adding to list all available nametemplates, since GET /nametemplate is already used, and I cannot break backwards compatibility. Hence I have added GET /nametemplates for this. Do you have a different suggestion?

Perhaps more discussion is needed in the [JIRA ticket](https://openmrs.atlassian.net/browse/RESTWS-938)?

dkayiwa commented on omod-2.0/src/test/resources/nameTemplate.xml:
```diff
@@ -1,17 +0,0 @@
-<org.openmrs.layout.name.NameTemplate>
```
Just to confirm, you removed this because it was not used anywhere. Correct?

dkayiwa commented on omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplatesController2_0.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * API endpoint to get a list of all available NameTemplates.
+ * <p>
+ * Unfortunately, there already exists a "/nametemplate" endpoint - see NameTemplateController2_0
```
Instead of creating a new end point simply to get a list of templates, i would rather that we fix the existing one, even if it means breaking the existing API.  Clients that are not ready to take advantage of this fix can keep using the previous version of the module and upgrade when they are ready to do the necessary changes.

xprl-gjf replied:
OK, I'm happy to do this. However, since that is a bigger change, I will cancel this PR and deliver in two parts:
* first, refactor existing `AddressTemplateController2_0` and `NameTemplateController2_0` to eliminate duplicate code.
* then, make the breaking changes to NameTemplate handling.

I did double check by raising a discussion point in Jira regarding breaking the API, but with no response. On that basis, will proceed.

xprl-gjf commented:
Superseded by #643 and with subsequent PR to come.
